PR - Gender Update

Kiyahitayika opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Gives the DNA's Gender state three values instead of two. 

Not sure why a binary state for gender existed in the DNA code since it could have easily been scripted otherwise. Now see the **trinary** state for gender - IPCs, rejoice!
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
This allows clones of tri-gendered species to properly carry their genders over to new bodies (e.g. Greys, Drask, etc.). It also makes it so tri-gendered species can store their gender _on_ the DNA. I mean, why wouldn't you want it there? That's where it is for bi-gendered species.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: GetUITriState(), SetUITriState()
tweak: Replace the GetUIState() -- only ever used for gender -- instances with GetUITriState() coding. Proofed for bi-gendered species.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kiyahitayika commented:
`HAS_GENDER` flag remains for Character Setup in order to select "Genderless".

Kiyahitayika commented:
Caught a bug/issue in my code. It was going to assign inconsistently-gendered body part sprites to Genderless folks (head, torso, groin mismatch). Fixed.

farie82 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,40 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+/datum/dna/proc/GetUITriState(block)
+	if(block <= 0)
+		return
+	else
```
Else not required here due to the early return at line 196

Kiyahitayika replied:
Oh! Sweet!

farie82 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,40 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+		return
+	else
+		var/val = GetUIValue(block)
+		if(val >= 1395 && val <= 2760) //female
```
Can make this a switch
```
switch(value)
  if(1 to 1395)
   ...
...
  else
   ...
```

Kiyahitayika replied:
Not sure if I did this correctly, I'm going to DM you on Discord

farie82 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,40 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+		var/val = GetUIValue(block)
+		if(val >= 1395 && val <= 2760) //female
+			return 0
+		if(val >= 1 && val <= 1395) //male
```
the comment makes no sense here since this is a generic helper

Kiyahitayika replied:
Oh! I put that in there because this will change how the values are assigned and we'll need to update the Wiki with the new Hexcodes should this get merged:
https://www.paradisestation.org/wiki/index.php/Guide_to_Genetics#Unique_Identifiers_.28UI.29

Instead of: 
Gender:
< 800 = Male
> 800 = Female

It will be:
Gender:
< 54F = Male
> 54F = Female = < 81C
> 81C Plural

farie82 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,40 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+			return 0
+		if(val >= 1 && val <= 1395) //male
+			return 1
+		if(val >= 2076 && val <= 4095) //plural
```
Same

farie82 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,40 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+	ASSERT(value <= 2)
+	var/val
+	switch(value)
+		if(0) //female
```
Same here for the comments

farie82 commented on code/game/dna/dna2_helpers.dm:
```diff
@@ -151,11 +150,13 @@
-			else
+		switch(dna.GetUITriState(DNA_UI_GENDER))
+			if(0)
+				H.change_gender(FEMALE,0)
```
spacing and `FALSE` instead of `0`. Boolean values should be set to either `TRUE` or `FALSE`. This improves readability.
Same for the ones below

Kiyahitayika replied:
Mmm. Gotcha!

farie82 commented on code/game/objects/items/weapons/dna_injector.dm:
```diff
@@ -43,7 +43,7 @@
 		return buf.dna.GetSEState(real_block)
 	else
-		return buf.dna.GetUIState(real_block)
+		return buf.dna.GetUITriState(real_block)
```
This will break a ton of the other UI data

Kiyahitayika replied:
**_HORF_**.

Kiyahitayika replied:
Somehow, someway, someone (me) thought this was related to gender.
👏 nice job me 👏

farie82 commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -1289,7 +1289,10 @@ GLOBAL_LIST_EMPTY(damage_icon_parts)
 	var/hulk = (HULK in mutations)
 	var/skeleton = (SKELETON in mutations)
-
+	var/g = dna.GetUITriState(DNA_UI_GENDER)
```
This will mean that every time this proc is called that the human will look different

Kiyahitayika replied:
:clique: :fingerguns: >__<;

farie82 commented on code/modules/surgery/organs/organ_icon.dm:
```diff
@@ -162,10 +162,13 @@ GLOBAL_LIST_EMPTY(limb_icon_cache)
+				if(1)
+					gender ="m"
+				else
+					gender = pick("f", "m")
```
Same as above

Kiyahitayika replied:
We don't have Genderless/Plural sprites. I'll default this to one value.

variableundefined commented:
@Kiyahitayika Avoid webedit and ensure the code are tested locally. Seeing 7 webedits that failed Travis in a row does not give us confidence the code will work as intended.

Kiyahitayika commented:
> Seeing 7 webedits that failed Travis in a row does not give us confidence the code will work as intended.

Maybe I'm new to Github! I didn't know how to webedit more than one file at a time and wasn't aware Travis was running tests. I thought it would run after changed each of the files. You learn something knew everyday.

Kiyahitayika commented:
Tested locally after web edits. Still works.

Changing species/gender via CMA - Admin:
![image](https://user-images.githubusercontent.com/69871346/94293965-150ce800-ff2d-11ea-94e2-3d8008cfc727.png)

Using Stable Mutagen to clone a Genderless/Plural Drask. The gender now transfers instead of defaulting to male:
![image](https://user-images.githubusercontent.com/69871346/94293976-1a6a3280-ff2d-11ea-96da-0ad00102c01b.png)

Using Stable Mutagen to clone a Female Grey. The gender now transfers instead of defaulting to male:
![image](https://user-images.githubusercontent.com/69871346/94295473-48e90d00-ff2f-11ea-87e9-9edf8eb47d1c.png)

(This will _**not**_ work for using Stable Mutagen on Vox because of how snowflaked the Vox code is. Neither did it work before this change. Fixing that is in the future and the problem is in the Vox coding, not the DNA Gender coding. This PR fixes the issue, however, for other races with Male/Female/Plural options.)

AffectedArc07 commented on code/game/dna/dna2.dm:
```diff
@@ -121,10 +120,12 @@ GLOBAL_LIST_EMPTY(bad_blocks)
-		SetUIState(DNA_UI_GENDER, character.gender!=MALE, 1)
-	else
-		SetUIState(DNA_UI_GENDER, pick(0,1), 1)
+	if(character.gender == MALE)
+		SetUITriState(DNA_UI_GENDER, 1, 1)
+	if(character.gender == FEMALE)
+		SetUITriState(DNA_UI_GENDER, 0, 1)
+	if(character.gender == PLURAL)
+		SetUITriState(DNA_UI_GENDER, 2, 1)
```
```suggestion
	switch(character.gender)
		if(MALE)
			SetUITriState(DNA_UI_GENDER, 1, 1)
		if(FEMALE)
			SetUITriState(DNA_UI_GENDER, 0, 1)
		if(PLURAL)
			SetUITriState(DNA_UI_GENDER, 2, 1)
```

AffectedArc07 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,38 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+	var/val = GetUIValue(block)
+	switch(val)
+		if(1395 to 2760)
+			return 0
```
What is 0,1,2 in this context? Magic numbers are less than good, so these should really be defined somewhere in `code/__DEFINES`, like so

```dm
#define WHATEVER_0_IS 0
#define WHATEVER_1_IS 1
#define WHATEVER_2_IS 2
```

Kiyahitayika replied:
Defined as DNA_GENDER_FEMALE, DNA_GENDER_FEMALE, DNA_GENDER_PLURAL.

AffectedArc07 commented on code/game/dna/dna2.dm:
```diff
@@ -189,6 +190,38 @@ GLOBAL_LIST_EMPTY(bad_blocks)
+	ASSERT(value <= 2)
+	var/val
+	switch(value)
+		if(0)
+			val = rand(1395, 2760)
+		if(1)
+			val = rand(1, 1395)
+		if(2) //plural
```
You'd then use `WHATEVER_1_IS` instead of `1` in all these if statements

AffectedArc07 commented on code/modules/surgery/organs/organ_icon.dm:
```diff
@@ -162,10 +162,13 @@ GLOBAL_LIST_EMPTY(limb_icon_cache)
+				if(1)
+					gender = "m"
+				else
+					gender = "f"
```
```suggestion
					gender = pick("f", "m")
```

Kiyahitayika replied:
![image](https://user-images.githubusercontent.com/69871346/94300214-ab91d700-ff36-11ea-9fcd-dc8bdc8c06b1.png)

Kiyahitayika replied:
Diona - same sprites
Drask - same sprites
Grey - different torso (the female one has breasts), different groin (female one is wider)
Vox - same sprites
Machine - same sprites

By "pick"ing gender each time, different sprites (for Grey) might be chosen depending on the organ. 

This proc is a getter, as per @farie82 's earlier comments, I'm concerned about having this proc ask for the organ icon file names and getting something like:

- Asks for torso icon, plural gender, picks from "f" and "m", sprite grabbed is "torso_f"
- Asks for groin icon for the same person, plural gender, picks from "f" and "m", sprite grabbed is "groin_m"

Right now, it doesn't matter so much since many tri-gendered species have the same sprites for "f" and "m", but if someone re-sprites them to be different, it might become noticeable.

Thoughts?

Kiyahitayika replied:
Since "f" is used as the default in line 162, I'm going to use "f".

AffectedArc07 reviewed (changes requested):
Just a few things

Kiyahitayika commented:
Changes made on local then re-commited.
Change requested was to replace the 0, 1, 2 format with #defines. #defines in /code/_DEFINES/dna.dm.

- 0 replaced with DNA_GENDER_FEMALE 
- 1 replaced with DNA_GENDER_MALE
- 2 replaced with DNA_GENDER_PLURAL

Tested again on local, still works.
![image](https://user-images.githubusercontent.com/69871346/94325351-ff6ae300-ff6b-11ea-9295-536fbefbd720.png)

Kiyahitayika commented:
Got our heads too deep in using this just for DNA_UI_GENDERr - returning the Set and Get procs to 0, 1, 2 - per @AffectedArc07 's request.

While in there, noticed some housekeeping items:
- reorganized the ranges so they flow numerically now
- changed the ranges so they don't overlap (line 205 having been the worst)
- removed the "else" statement from line 222 since we're using ASSERTS. There shouldn't be a value outside the range of 0 to 2 for this.
